base/sparse/linalg.jl

```diff
@@ -499,7 +499,196 @@ function norm(A::SparseMatrixCSC,p::Real=2)
     throw(ArgumentError("invalid p-norm p=$p. Valid: 1, Inf"))
 end
 
-# TODO
+# TODO rank
+
+# cond
+function cond(A::SparseMatrixCSC, p::Real=1)
+    if p == 1
+        normA = norm(A, 1)
+        normAinv = normestinv(A)
+        return normA * normAinv
+    elseif p == Inf
+        normA = norm(A, Inf)
+        normAinv = normestinv(A')
+        return normA * normAinv
+    elseif p == 2
+        throw(ArgumentError("2-norm condition number is not implemented for sparse matrices, try cond(full(A), 2) instead"))
+    else
+        throw(ArgumentError("second argment must be either 1 or Inf, got $p"))
+    end
+end
+
+function normestinv{T}(A::SparseMatrixCSC{T}, t::Integer = 2)
+
+    maxiter = 5
+
+    # Check the input
+    sizeA = size(A)
+    n = max(sizeA[1],sizeA[2])
+    F = factorize(A)
+    t = min(t,n)
+    if ~isreal(t) || length(t) > 1 || t <= 0
+        throw(ArgumentError("number of blocks must be a positive integer"))
+    end
+    if t > n
+        throw(ArgumentError("number of blocks must not be greater than $n"))
+    end
+    old_est = 0
+    ind = Array(Int64, n)
+    ind_hist = Array(Int64, maxiter * t)
+    S = zeros(T, n, t)
+
+    # Generate the block matrix
+    h = zeros(real(T), n)
+    X = Array(T, n, t)
+    for i = 1:n
+        X[i, 1] = 1
+    end
+    for j = 2:t
+        repeated = true
+        while repeated
+            for i = 1:n
+                X[i,j] = rand()>=0.5?1:-1
+            end
+            repeated = false
+            yaux = X[1:n,j]' * X[1:n,1:j-1]
+            for i = 1:j-1
+                if abs(yaux[i]) == n
+                    repeated = true
+                    break
+                end
+            end
+        end
+    end
+    scale!(X, 1./n)
+
+    iter = 0
+    est_old = 0
+    est = 0
```

Comment:
This causes the return type to be a `Union{Int64,Float64}` and there are several other places where the type of the variable changes during the function. You should call 
```jl
code_warntype(SparseMatrix.normestinv, (SparseMatrixCSC{Float64,Int64}, Int))
```
and see how many of the red words you can get rid of. The return type from `factorize` cannot be made unique, but it shouldn't infect the rest of the code. If it does, we'll have to add some `::T`s, but don't do this initially.